File: examples/node/ts/ackPayloads.ts
/**
 * Simple example of using the nRF24L01 to transmit and
 * retrieve automatic acknowledgment packets with customized payloads.
 *
 * This example is meant to be run on 2 separate nRF24L01 transceivers.
 *
 * See documentation at https://nRF24.github.io/rf24-rs
 */
import * as readline from "readline/promises";
import * as timer from "timers/promises";
import { RF24, PaLevel } from "@rf24/rf24";

const io = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

export class App {
  radio: RF24;
  counter: number;

  /**
   * @param radioNumber The number (0 or 1) that identifies the which radio is used in
   * this example that requires 2 radios.
   */
  constructor(radioNumber: number) {
    // The radio's CE Pin uses a GPIO number.
    const cePin = 22; // for GPIO22

    // The radio's CSN Pin corresponds the SPI bus's CS pin (aka CE pin).
    // On Linux, consider the device path `/dev/spidev<a>.<b>`:
    //   - `<a>` is the SPI bus number (defaults to `0`)
    //   - `<b>` is the CSN pin (must be unique for each device on the same SPI bus)
    const csnPin = 0; // aka CE0 for SPI bus 0 (/dev/spidev0.0)

    // create a radio object for the specified hardware config:
    this.radio = new RF24(cePin, csnPin);

    // initialize the nRF24L01 on the spi bus
    this.radio.begin();

    // For this example, we will use different addresses
    // An address needs to be a buffer object (bytearray)
    const address = [Buffer.from("1Node"), Buffer.from("2Node")];

    //set TX address of RX node into the TX pipe
    this.radio.asTx(address[radioNumber]); // always uses pipe 0

    // set RX address of TX node into an RX pipe
    this.radio.openRxPipe(1, address[1 - radioNumber]); // using pipe 1

    // set the Power Amplifier level to -12 dBm since this test example is
    // usually run with nRF24L01 transceivers in close proximity of each other
    this.radio.paLevel = PaLevel.Low; // PaLevel.Max is default

    this.radio.ackPayloads = true;
    this.radio.dynamicPayloads = true;

    this.counter = 0;
  }

  /**
   * The transmitting node's behavior.
   * @param count The number of payloads to send
   */
  async tx(count?: number) {
    this.radio.asTx();
    // we'll use a DataView object to store our string and number into a bytearray buffer
    const outgoing = Buffer.from("Hello \0.");
    for (let i = 0; i < (count || 5); i++) {
      outgoing.writeUint8(this.counter, 7);
      const start = process.hrtime.bigint();
      const result = this.radio.send(outgoing);
      const end = process.hrtime.bigint();
      if (result) {
        const elapsed = (end - start) / BigInt(1000);
        process.stdout.write(
          `Transmission successful! Time to Transmit: ${elapsed} us. Sent: ` +
            `${outgoing.subarray(0, 6).toString("utf-8")}${this.counter} `,
        );
        this.counter += 1;
        if (this.radio.available()) {
          const incoming = this.radio.read();
          const counter = incoming.readUint8(7);
          console.log(
            ` Received: ${incoming.subarray(0, 6).toString("utf-8")}${counter}`,
          );
        } else {
          console.log("Received an empty ACK packet");
        }
      } else {
        console.log("Transmission failed or timed out!");
      }
      await timer.setTimeout(1000);
    }

    // recommended behavior is to keep in TX mode while idle
    this.radio.asTx(); // enter inactive TX mode
  }

  /**
   * The receiving node's behavior.
   * @param duration The timeout duration (in seconds) to listen after receiving a payload.
   */
  rx(duration?: number) {
    this.radio.asRx();
    // we'll use a DataView object to store our string and number into a bytearray buffer
    const outgoing = Buffer.from("World \0.");
    outgoing.writeUint8(this.counter, 7);
    this.radio.writeAckPayload(1, outgoing);
    let timeout = Date.now() + (duration || 6) * 1000;
    while (Date.now() < timeout) {
      const hasRx = this.radio.availablePipe();
      if (hasRx.available) {
        const incoming = this.radio.read();
        const counter = incoming.readUint8(7);
        console.log(
          `Received ${incoming.length} bytes on pipe ${hasRx.pipe}: ` +
            `${incoming.subarray(0, 6).toString("utf-8")}${counter} Sent: ` +
            `${outgoing.subarray(0, 6).toString("utf-8")}${this.counter}`,
        );
        this.counter = counter;
        outgoing.writeUint8(counter + 1, 7);
        this.radio.writeAckPayload(1, outgoing);
        timeout = Date.now() + (duration || 6) * 1000;
      }
    }

    // recommended behavior is to keep in TX mode while idle
    this.radio.asTx(); // enter inactive TX mode
    // `asTx()` also flushes any unused ACK payloads in the TX FIFO
    // when ACK payloads are enabled
  }

  /**
   * This function prompts the user and performs the specified role for the radio.
   */
  async setRole() {
    const prompt =
      "*** Enter 'T' to transmit\n" +
      "*** Enter 'R' to receive\n" +
      "*** Enter 'Q' to quit\n";
    io.resume();
    const input = (await io.question(prompt)).split(" ");
    io.pause();
    const role = input.shift() || "?";
    const params = Array<number>();
    for (let i = 0; i < input.length; ++i) {
      params.push(Number(input[i]));
    }
    switch (role.charAt(0).toLowerCase()) {
      case "t":
        await this.tx(...params);
        return true;
      case "r":
        this.rx(...params);
        return true;
      default:
        console.log(`'${role.charAt(0)}' is an unrecognized input`);
        return true;
      case "q":
        this.radio.powerDown();
        return false;
    }
  }
}

export async function main() {
  console.log(module.filename);

  io.resume();
  // to use different addresses on a pair of radios, we need a variable to
  // uniquely identify which address this radio will use to transmit
  // 0 uses address[0] to transmit, 1 uses address[1] to transmit
  const radioNumber = Number(
    (await io.question(
      "Which radio is this? Enter '1' or '0' (default is '0') ",
    )) == "1",
  );
  console.log(`radioNumber is ${radioNumber}`);
  io.pause();

  const app = new App(radioNumber);
  while (await app.setRole());
  io.close();
}

main();

File: examples/rust/src/bin/ack_payloads.rs
#![no_std]

use core::time::Duration;

use anyhow::{anyhow, Result};
use embedded_hal::delay::DelayNs;
use rf24::{
    radio::{prelude::*, RF24},
    PaLevel,
};
#[cfg(feature = "linux")]
use rf24_rs_examples::linux::{
    print, println, BoardHardware, CdevPin as DigitalOutImpl, Delay as DelayImpl,
    SpidevDevice as SpiImpl,
};
#[cfg(feature = "linux")]
extern crate std;
#[cfg(feature = "linux")]
use std::{
    borrow::ToOwned,
    io::Write,
    string::{String, ToString},
    time::Instant,
};

/// A struct to drive our example app
struct App {
    /// Any platform-specific functionality is abstracted into this object.
    #[allow(dead_code, reason = "keep board's peripheral objects alive")]
    board: BoardHardware,
    /// Our instantiated RF24 object.
    radio: RF24<SpiImpl, DigitalOutImpl, DelayImpl>,
    /// We will be using a incrementing integer value as part of our payloads.
    counter: u8,
}

impl App {
    pub fn new() -> Result<Self> {
        // instantiate a hardware peripherals on the board
        let mut board = BoardHardware::default()?;

        // instantiate radio object using board's hardware
        let radio = RF24::new(
            board.default_ce_pin()?,
            BoardHardware::default_spi_device()?,
            DelayImpl,
        );
        Ok(Self {
            board,
            radio,
            counter: 0,
        })
    }

    /// Setup the radio for this example.
    ///
    /// This will initialize and configure the [`App::radio`] object.
    pub fn setup(&mut self, radio_number: u8) -> Result<()> {
        // initialize the radio hardware
        self.radio.init().map_err(|e| anyhow!("{e:?}"))?;

        // defaults to PaLevel::Max. Use PaLevel::Low for PA/LNA testing
        self.radio
            .set_pa_level(PaLevel::Low)
            .map_err(|e| anyhow!("{e:?}"))?;

        // we'll be using a automatic ACK payloads, so enable the dynamic payload length feature
        self.radio
            .set_dynamic_payloads(true)
            .map_err(|e| anyhow!("{e:?}"))?;
        // enable ACK packet payloads
        self.radio
            .set_ack_payloads(true)
            .map_err(|e| anyhow!("{e:?}"))?;

        let address = [b"1Node", b"2Node"];
        self.radio
            .open_tx_pipe(address[radio_number as usize])
            .map_err(|e| anyhow!("{e:?}"))?;
        self.radio
            .open_rx_pipe(1, address[1 - radio_number as usize])
            .map_err(|e| anyhow!("{e:?}"))?;
        Ok(())
    }

    /// The TX role.
    ///
    /// Uses the [`App::radio`] as a transmitter.
    pub fn tx(&mut self, count: u8) -> Result<()> {
        // put radio into TX mode
        self.radio.as_tx().map_err(|e| anyhow!("{e:?}"))?;

        // declare our outgoing payload.
        // `\x00` is null terminator for the string portion.
        // `0` is placeholder for our counter value.
        let mut outgoing_payload = b"Hello \x000".to_owned();
        let mut remaining = count;
        while remaining > 0 {
            outgoing_payload[7] = self.counter;
            let start = Instant::now();
            let result = self
                .radio
                .send(&outgoing_payload, false)
                .map_err(|e| anyhow!("{e:?}"))?;
            let end = Instant::now();
            if result {
                // succeeded
                print!(
                    "Transmission successful! Time to Transmit: {} us. Sent: {}{} Received: ",
                    end.saturating_duration_since(start).as_micros(),
                    String::from_utf8_lossy(&outgoing_payload[0..6]),
                    self.counter,
                );
                self.counter += 1;
                if self.radio.available().map_err(|e| anyhow!("{e:?}"))? {
                    let mut incoming_payload = [0u8; 8];
                    self.radio
                        .read(&mut incoming_payload, None)
                        .map_err(|e| anyhow!("{e:?}"))?;
                    println!(
                        "{}{}",
                        String::from_utf8_lossy(&incoming_payload[0..6]),
                        incoming_payload[7],
                    );
                } else {
                    println!("An empty ACK payload");
                }
            } else {
                // failed
                println!("Transmission failed or timed out");
            }
            remaining -= 1;
            DelayImpl.delay_ms(1000);
        }
        Ok(())
    }

    /// The RX role.
    ///
    /// Uses the [`App::radio`] as a receiver.
    pub fn rx(&mut self, timeout: u8) -> Result<()> {
        // put radio into active RX mode
        self.radio.as_rx().map_err(|e| anyhow!("{e:?}"))?;

        // declare our outgoing payload
        // `\x00` is the null terminator for the string portion
        // `0` is the placeholder for our counter value
        let mut outgoing_payload = b"World \x000".to_owned();
        outgoing_payload[7] = self.counter;
        // load ACK for first response
        self.radio
            .write_ack_payload(1, &outgoing_payload)
            .map_err(|e| anyhow!("{e:?}"))?;

        let mut end_time = Instant::now() + Duration::from_secs(timeout as u64);
        while Instant::now() < end_time {
            let mut pipe = 15u8;
            if self
                .radio
                .available_pipe(&mut pipe)
                .map_err(|e| anyhow!("{e:?}"))?
            {
                let mut incoming_payload = [0u8; 8];
                let len = self
                    .radio
                    .read(&mut incoming_payload, None)
                    .map_err(|e| anyhow!("{e:?}"))?;
                self.counter = incoming_payload[7];
                // print pipe number and payload length and payload
                println!(
                    "Received {len} bytes on pipe {pipe}: {}{} Sent: {}{}",
                    String::from_utf8_lossy(&incoming_payload[0..6]),
                    self.counter,
                    String::from_utf8_lossy(&outgoing_payload[0..6]),
                    outgoing_payload[7],
                );
                // reset timeout
                end_time = Instant::now() + Duration::from_secs(timeout as u64);

                // increment counter
                self.counter += 1;
                outgoing_payload[7] = self.counter;
                // load new ACK payload for next response
                self.radio
                    .write_ack_payload(1, &outgoing_payload)
                    .map_err(|e| anyhow!("{e:?}"))?;
            }
        }

        // It is highly recommended to keep the radio idling in an inactive TX mode
        self.radio.as_tx().map_err(|e| anyhow!("{e:?}"))?;
        // as_tx() will also flush any remaining ACK payloads from the radio's TX FIFO.
        Ok(())
    }

    pub fn set_role(&mut self) -> Result<bool> {
        let prompt = "*** Enter 'R' for receiver role.\n\
        *** Enter 'T' for transmitter role.\n\
        *** Enter 'Q' to quit example.";
        println!("{prompt}");
        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
        let mut inputs = input.trim().split(' ');
        let role = inputs
            .next()
            .map(|v| v.to_uppercase())
            .unwrap_or("?".to_string());
        if role.starts_with('T') {
            let count = inputs
                .next()
                .and_then(|v| v.parse::<u8>().ok())
                .unwrap_or(5);
            self.tx(count)?;
            return Ok(true);
        } else if role.starts_with('R') {
            let timeout = inputs
                .next()
                .and_then(|v| v.parse::<u8>().ok())
                .unwrap_or(6);
            self.rx(timeout)?;
            return Ok(true);
        } else if role.starts_with('Q') {
            self.radio.power_down().map_err(|e| anyhow!("{e:?}"))?;
            return Ok(false);
        }
        println!("{role} is an unrecognized input. Please try again.");
        Ok(true)
    }
}

fn main() -> Result<()> {
    let mut app = App::new()?;
    let mut input = String::new();
    print!("Which radio is this? Enter '0' or '1'. Defaults to '0' ");
    std::io::stdout().flush()?;
    std::io::stdin().read_line(&mut input)?;
    let radio_number = input
        .trim()
        .chars()
        .next()
        .map(|c| if c == '1' { 1 } else { 0 })
        .unwrap_or_default();
    app.setup(radio_number)?;
    while app.set_role()? {}
    Ok(())
}
